arraykit.c

```diff
@@ -129,6 +130,62 @@ AK_ResolveDTypeIter(PyObject *dtypes)
     return resolved;
 }
 
+// Numpy implementation: https://github.com/numpy/numpy/blob/a14c41264855e44ebd6187d7541b5b8d59bb32cb/numpy/core/src/multiarray/methods.c#L1557
+PyObject*
+AK_ArrayCopy(PyObject *array, PyObject *memo)
+{
+    PyObject *id = PyLong_FromVoidPtr(array);
+    PyObject *found = PyDict_GetItem(memo, id);
+    if (found) { // found will be NULL if not in dict
+        Py_DECREF(id);
+        Py_INCREF(found); // got a borrowed ref
+        return found;
+    }
+
+    // if dtype is object, call deepcopy with memo
+    PyObject *array_new;
+    PyArray_Descr *dtype = PyArray_DESCR((PyArrayObject *)array); // borrowed ref
+
+    if PyDataType_ISOBJECT(dtype)
+    {
+        PyObject *copy = PyImport_ImportModule("copy");
+        if (!copy) {
+            Py_DECREF(id);
+            return NULL;
+        }
+        PyObject *deepcopy = PyObject_GetAttrString(copy, "deepcopy");
+        Py_DECREF(copy);
+        if (!deepcopy) {
+            Py_DECREF(id);
+            return NULL;
+        }
+        array_new = PyObject_CallFunctionObjArgs(deepcopy, array, memo, NULL);
+        Py_DECREF(deepcopy);
+        if (!array_new) {
+            Py_DECREF(id);
+            return NULL;
+        }
+    }
+    else {
+        Py_INCREF(dtype); // PyArray_FromArray steals a reference
+        array_new = PyArray_FromArray(
+                (PyArrayObject*) array,
+                dtype,
+                NPY_ARRAY_ENSURECOPY);
+        if (PyDict_SetItem(memo, id, array_new))
+        {
+            Py_DECREF(array_new);
+            Py_DECREF(id);
+            return NULL;
+        }
+    }
+    // set immutable
+    PyArray_CLEARFLAGS((PyArrayObject *)array_new, NPY_ARRAY_WRITEABLE);
+    Py_DECREF(id);
+    return array_new;
```

Comment:
Now you can `goto error;`!
```suggestion
    return array_new;
error:
    Py_DECREF(id);
    return NULL;
```

Reply:
Well, it had to happen eventually: my first `goto`. Sorry Dijkstra! But I do so the better reuse.